Sources/Basics/AuthorizationProvider.swift

```diff
@@ -24,3 +42,224 @@ extension AuthorizationProvider {
+        }
+        let machine = TSCUtility.Netrc.Machine(name: machineName, login: user, password: password)
+        
+        var machines = [TSCUtility.Netrc.Machine]()
+        var hasExisting = false
+        
+        self.machines.forEach {
+            if $0.name.lowercased() != machineName {
+                machines.append($0)
+            } else if !hasExisting {
+                // Update existing entry and retain one copy only
+                machines.append(machine)
+                hasExisting = true
+            }
+        }
+        
+        // New entry
+        if !hasExisting {
+            machines.append(machine)
+        }
+        
+        do {
+            try self.saveToDisk(machines: machines)
+            // At this point the netrc file should exist and non-empty
+            guard let netrc = try Self.loadFromDisk(path: self.path) else {
+                throw AuthorizationProviderError.unexpectedError("Failed to update netrc file at \(self.path)")
+            }
+            self.underlying = netrc
+            callback(.success(()))
+        } catch {
+            callback(.failure(AuthorizationProviderError.unexpectedError("Failed to update netrc file at \(self.path): \(error)")))
+        }
+    }
+
+    public func authentication(for url: Foundation.URL) -> (user: String, password: String)? {
+        self.machine(for: url).map { (user: $0.login, password: $0.password) }
+    }
+    
+    private func machineName(for url: Foundation.URL) -> String? {
+        url.host?.lowercased()
+    }
+
+    private func machine(for url: Foundation.URL) -> TSCUtility.Netrc.Machine? {
+        if let machineName = self.machineName(for: url), let machine = self.machines.first(where: { $0.name.lowercased() == machineName }) {
+            return machine
+        }
+        if let machine = self.machines.first(where: { $0.isDefault }) {
+            return machine
+        }
+        return .none
+    }
+    
+    private func saveToDisk(machines: [TSCUtility.Netrc.Machine]) throws {
+        try self.fileSystem.withLock(on: self.path, type: .exclusive) {
+            try self.fileSystem.writeFileContents(self.path) { stream in
+                machines.forEach {
+                    stream.write("machine \($0.name) login \($0.login) password \($0.password)\n")
+                }
+            }
+        }
+    }
+    
+    private static func loadFromDisk(path: AbsolutePath) throws -> TSCUtility.Netrc? {
+        do {
+            return try TSCUtility.Netrc.load(fromFileAtPath: path).get()
+        } catch {
+            switch error {
+            case Netrc.Error.fileNotFound, Netrc.Error.machineNotFound:
+                // These are recoverable errors. We will just create the file and append entry to it.
+                return nil
+            default:
+                throw error
+            }
+        }
+    }
+}
+
+// MARK: - Keychain
+
+#if os(macOS)
```

Comment:
we can probably expand this list to more Darwin based platforms just in case. cc @abertelrud

Reply:
Maybe we can use `canImport` to check if the Security framework is available, instead of making it a platform-based check

Reply:
Changing to use `canImport` per Boris' suggestion